modules/plugin/shapefile/src/test/java/org/geotools/data/shapefile/IndexManagerTest.java

```diff
@@ -0,0 +1,228 @@
+    public void testCreateSpatialIndex_Concurrency_runs_once() throws Exception {
+        deleteQixFile();
+        IndexManager indexManager = new IndexManager(shpFiles, mockDataStore);
+
+        final int threadCount = 8;
+        final int taskCount = 16;
+        final boolean force = false;
+        List<Boolean> builds =
+                buildSpatialIndexConcurrently(indexManager, threadCount, taskCount, force);
+
+        long buildCount = builds.stream().filter(Boolean::booleanValue).count();
+
+        String msg =
+                String.format(
+                        "Expected only 1 true result from %d concurrent createSpatialIndex calls",
+                        taskCount);
+        assertEquals(msg, 1, buildCount);
+    }
+
+    @Test(timeout = 10_000) // timeout at 10s to fail in case of deadlock
+    public void testCreateSpatialIndex_Concurrency_with_force_run_sequentially() throws Exception {
+        deleteQixFile();
+        final AtomicInteger concurrentBuilds = new AtomicInteger();
+        final AtomicInteger maxConcurrentBuilds = new AtomicInteger();
+        IndexManager indexManager =
+                new IndexManager(shpFiles, mockDataStore) {
+                    @Override
+                    protected void doCreateSpatialIndex() throws Exception {
+                        int concurrency = concurrentBuilds.incrementAndGet();
+                        super.doCreateSpatialIndex();
+                        concurrentBuilds.decrementAndGet();
+                        maxConcurrentBuilds.set(Math.max(maxConcurrentBuilds.get(), concurrency));
+                    }
+                };
+
+        final int threadCount = 8;
+        final int taskCount = 16;
+        final boolean force = true;
+        List<Boolean> builds =
+                buildSpatialIndexConcurrently(indexManager, threadCount, taskCount, force);
+
+        long buildCount = builds.stream().filter(Boolean::booleanValue).count();
+
+        String msg =
+                String.format(
+                        "Expected only %d true results from %d concurrent createSpatialIndex calls with force == true",
+                        taskCount, taskCount);
+        assertEquals(msg, taskCount, buildCount);
+        assertEquals(1, maxConcurrentBuilds.get());
+    }
+
+    private List<Boolean> buildSpatialIndexConcurrently(
+            IndexManager indexManager,
+            final int threadCount,
+            final int taskCount,
+            final boolean force)
+            throws Exception {
+        final ExecutorService executor = Executors.newFixedThreadPool(threadCount);
+        List<Callable<Boolean>> tasks =
+                IntStream.range(0, taskCount)
+                        .mapToObj(
+                                i ->
+                                        (Callable<Boolean>)
+                                                () -> indexManager.createSpatialIndex(force))
+                        .collect(Collectors.toList());
```

Comment:
An old style for loop ends up taking less space and requires no casts, be mindful of the 80 columns limit.

```suggestion
        List<Callable<Boolean>> tasks  = new ArrayList(); 
        for(int i = 0; i < taskCount; i++) 
           tasks.add(indexManager.createSpatialIndex(force));
```

Just 3 lines ;-) (yes, curly braces are not mandatory in Google AOSP codingin conventions)

Reply:
https://github.com/geotools/geotools/pull/3527/files#diff-cca0c8a31b2bd3b8cf2a2473aae2301caf3f80def759a5d36a2d82b3cc4e319bR196-R211